src/Libraries/Microsoft.Extensions.AI.Abstractions/Embeddings/BinaryEmbedding.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Represents an embedding composed of a bit vector.</summary>
+public sealed class BinaryEmbedding : Embedding
+{
+    /// <summary>The embedding vector this embedding represents.</summary>
+    private BitArray _vector;
+
+    /// <summary>Initializes a new instance of the <see cref="BinaryEmbedding"/> class with the embedding vector.</summary>
+    /// <param name="vector">The embedding vector this embedding represents.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="vector"/> is <see langword="null"/>.</exception>
+    public BinaryEmbedding(BitArray vector)
+    {
+        _vector = Throw.IfNull(vector);
+    }
+
+    /// <summary>Gets or sets the embedding vector this embedding represents.</summary>
+    [JsonConverter(typeof(VectorConverter))]
+    public BitArray Vector
+    {
+        get => _vector;
+        set => _vector = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    [JsonIgnore]
+    public override int Dimensions => _vector.Length;
+
+    /// <summary>Provides a <see cref="JsonConverter{BitArray}"/> for serializing <see cref="BitArray"/> instances.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class VectorConverter : JsonConverter<BitArray>
+    {
+        /// <inheritdoc/>
+        public override BitArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            _ = Throw.IfNull(typeToConvert);
+            _ = Throw.IfNull(options);
+
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw new JsonException("Expected string property.");
+            }
+
+            ReadOnlySpan<byte> utf8;
+            byte[]? tmpArray = null;
+            if (!reader.HasValueSequence && !reader.ValueIsEscaped)
+            {
+                utf8 = reader.ValueSpan;
+            }
+            else
+            {
+                if (reader.HasValueSequence)
+                {
+                    long sequenceLength = reader.ValueSequence.Length;
+                    if (sequenceLength > int.MaxValue)
+                    {
+                        throw new JsonException("Value is too long.");
+                    }
+
+                    tmpArray = ArrayPool<byte>.Shared.Rent((int)sequenceLength);
```

Comment:
Consider using a stack allocated buffer for small lengths, e.g.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs#L244-L246

Reply:
Is it worth it? The 99.999% case should be deserializing strings we created, which means they shouldn't have any escape sequences in them, and it sounds like it'll basically never be a `ReadOnlySequence<byte>` if I understood you correctly.